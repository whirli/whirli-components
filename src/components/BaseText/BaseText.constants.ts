export const PropKeys = {
  COLOR: 'color',
  TAG: 'tag',
  SIZE: 'size',
  FONT: 'font',
  TRANSFORM: 'transform',
  LETTER_SPACING: 'letterSpacing',
  LINE_HEIGHT: 'lineHeight',
  ALIGN: 'align',
};

export const PropValues = {
  [PropKeys.COLOR]: ['inherit'],
  [PropKeys.TAG]: [
    'a',
    'nuxt-link',
    'router-link',
    'address',
    'time',
    'span',
    'strong',
    'div',
    'p',
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
    'address',
    'li',
    'blockquote',
    'dt',
    'dd',
    'figcaption',
  ],
  [PropKeys.SIZE]: ['3xs', '2xs', 'xs', 'sm', 'base', 'md', 'lg', 'xl', '2xl', '3xl', '4xl', '5xl'],
  [PropKeys.FONT]: ['default', 'bold', 'mono'],
  [PropKeys.TRANSFORM]: ['default', 'uppercase'],
  [PropKeys.LETTER_SPACING]: ['default', 'md', 'lg', 'xl'],
  [PropKeys.LINE_HEIGHT]: ['3xs', '2xs', 'xs', 'sm', 'base', 'md', 'lg', 'xl', '2xl', '3xl', '4xl', '5xl'],
  [PropKeys.ALIGN]: ['default', 'center', 'right'],
};
