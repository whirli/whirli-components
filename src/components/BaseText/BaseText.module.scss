@import '../../styles/Breakpoints.scss';
@import '../../styles/Helpers.scss';

@font-face {
    font-family: "Maax";
    src: url('~@whirli-components/src/assets/fonts/Maax.woff2') format("woff2");
}

@font-face {
    font-family: "MaaxBold";
    src: url('~@whirli-components/src/assets/fonts/MaaxBold.woff2') format("woff2");
}

@font-face {
    font-family: "MaaxMono";
    src: url('~@whirli-components/src/assets/fonts/MaaxMonoRegular.woff2') format("woff2");
}

// default styles
.text {}

.transform-default {
    text-transform: none;
}

.transform-uppercase {
    text-transform: uppercase;
}

.font-sans {
    font-family: 'Maax', Tahoma, Geneva, "Helvetica Neue", Helvetica, Arial, sans-serif;
}

.font-bold {
    font-family: 'MaaxBold', Tahoma, Geneva, "Helvetica Neue", Helvetica, Arial, sans-serif;
    font-weight: bold;
}

.font-mono {
    font-family: 'MaaxMono', "Courier New", Courier, "Lucida Sans Typewriter", "Lucida Typewriter", monospace;
}


// breakpoint styles
$packageBreakpointStyles: (
    color-inherit: (
        color: inherit,
    ),
    size-3xs: (
        font-size: toRems(10),
    ),
    size-2xs: (
        font-size: toRems(12),
    ),
    size-xs: (
        font-size: toRems(13),
    ),
    size-sm: (
        font-size: toRems(14),
    ),
    size-default: (
        font-size: toRems(16),
    ),
    size-md: (
        font-size: toRems(18),
    ),
    size-lg: (
        font-size: toRems(20),
    ),
    size-xl: (
        font-size: toRems(24),
    ),
    size-2xl: (
        font-size: toRems(28),
    ),
    size-3xl: (
        font-size: toRems(32),
    ),
    size-4xl: (
        font-size: toRems(40),
    ),
    size-5xl: (
        font-size: toRems(48),
    ),
    line-height-3xs: (
        font-size: toRems(16),
    ),
    line-height-2xs: (
        font-size: toRems(18),
    ),
    line-height-xs: (
        font-size: toRems(20),
    ),
    line-height-sm: (
        font-size: toRems(22),
    ),
    line-height-default: (
        font-size: toRems(24),
    ),
    line-height-md: (
        font-size: toRems(28),
    ),
    line-height-lg: (
        font-size: toRems(32),
    ),
    line-height-xl: (
        font-size: toRems(36),
    ),
    line-height-2xl: (
        font-size: toRems(40),
    ),
    line-height-3xl: (
        font-size: toRems(42),
    ),
    line-height-4xl: (
        font-size: toRems(48),
    ),
    line-height-5xl: (
        font-size: toRems(56),
    ),
    letter-spacing-default: (
        font-size: toRems(0),
    ),
    letter-spacing-md: (
        font-size: toRems(1),
    ),
    letter-spacing-lg: (
        font-size: toRems(1.2),
    ),
    letter-spacing-xl: (
        font-size: toRems(1.5),
    ),
    align-default: (
        text-align: left,
    ),
    align-center: (
        text-align: center,
    ),
    align-right: (
        text-align: right,
    )
);