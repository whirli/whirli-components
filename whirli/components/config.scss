@import '@whirli-components/styles/Helpers.scss';

$breakpointPrefix: BREAKPOINT_;

$breakpoints: (
    tablet: (
        prefix: t,
        min-width: 672px,
    ),
    small_desk: (
        prefix: sd,
        min-width: 1056px
    ),
    large_desk: (
        prefix: ld,
        min-width: 1312px
    ),
    huge_desk: (
        prefix: hd,
        min-width: 1584px
    ),
);

:export {
    prefix: $breakpointPrefix;
    #{$breakpointPrefix}default: 0px;
    @each $name, $value in $breakpoints {
        #{$breakpointPrefix}#{$name}: map-deep-get($breakpoints, #{$name}, 'min-width')
    }
}