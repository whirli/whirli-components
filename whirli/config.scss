@import '@whirli-components/src/styles/Helpers.scss';

$breakpoints: (
    tablet: (
        prefix: t,
        min-width: 48rem,
    ),
    desktop: (
        prefix: d,
        min-width: 80rem,
    )
);

:export {
    @each $name, $value in $breakpoints {
        BREAKPOINT_#{to-upper-case(#{$name})}: map-deep-get($breakpoints, #{$name}, 'min-width')
    }
}