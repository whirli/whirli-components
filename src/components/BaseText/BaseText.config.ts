// @ts-ignore
import { PropType } from '@composition';
import * as PropHelpers from '@whirli-components/helpers/props';
import { ComponentStyles, ComponentProps } from '@whirli-components/@types/components';

// Local config
import config from '@whirli-local/components/BaseText/BaseText.config';

// Package config
import { PropKeys, PropValues } from './BaseText.constants';
import type {
  PropColor,
  PropSize,
  PropTag,
  PropFont,
  PropTransform,
  PropLetterSpacing,
  PropLineHeight,
  PropAlign,
} from './BaseText.types';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
  [PropKeys.COLOR]: {
    generateBreakpoints: true,
    classes: {
      default: 'color-inherit',
      ...config.styles[PropKeys.COLOR]?.classes,
    } as Record<PropColor, string>,
  },
  [PropKeys.SIZE]: {
    generateBreakpoints: true,
    classes: {
      '3xs': 'size-3xs',
      '2xs': 'size-2xs',
      xs: 'size-xs',
      sm: 'size-sm',
      default: 'size-default',
      md: 'size-md',
      lg: 'size-lg',
      xl: 'size-xl',
      '2xl': 'size-2xl',
      '3xl': 'size-3xl',
      '4xl': 'size-4xl',
      '5xl': 'size-5xl',
      ...config.styles[PropKeys.SIZE]?.classes,
    } as Record<PropSize, string>,
  },
  [PropKeys.FONT]: {
    generateBreakpoints: true,
    classes: {
      default: 'font-sans',
      bold: 'font-bold',
      mono: 'font-mono',
      ...config.styles[PropKeys.FONT]?.classes,
    } as Record<PropFont, string>,
  },
  [PropKeys.TRANSFORM]: {
    classes: {
      default: 'transform-default',
      uppercase: 'transform-uppercase',
      ...config.styles[PropKeys.TRANSFORM]?.classes,
    } as Record<PropTransform, string>,
  },
  [PropKeys.LETTER_SPACING]: {
    generateBreakpoints: true,
    classes: {
      default: 'letter-spacing-default',
      md: 'letter-spacing-md',
      lg: 'letter-spacing-lg',
      xl: 'letter-spacing-xl',
      ...config.styles[PropKeys.LETTER_SPACING]?.classes,
    } as Record<PropLetterSpacing, string>,
  },
  [PropKeys.LINE_HEIGHT]: {
    generateBreakpoints: true,
    classes: {
      '3xs': 'line-height-3xs',
      '2xs': 'line-height-2xs',
      xs: 'line-height-xs',
      sm: 'line-height-sm',
      default: 'line-height-default',
      md: 'line-height-md',
      lg: 'line-height-lg',
      xl: 'line-height-xl',
      '2xl': 'line-height-2xl',
      '3xl': 'line-height-3xl',
      '4xl': 'line-height-4xl',
      '5xl': 'line-height-5xl',
      ...config.styles[PropKeys.LINE_HEIGHT]?.classes,
    } as Record<PropLineHeight, string>,
  },
  [PropKeys.ALIGN]: {
    generateBreakpoints: true,
    classes: {
      default: 'align-default',
      center: 'align-center',
      right: 'align-right',
      ...config.styles[PropKeys.ALIGN]?.classes,
    } as Record<PropAlign, string>,
  },
};

export const ConfigProps: ComponentProps = {
  [PropKeys.COLOR]: {
    required: false,
    type: [String, Object] as PropType<PropColor>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.COLOR),
  },
  [PropKeys.TAG]: {
    required: false,
    type: String as PropType<PropTag>,
    validator: (value: PropTag): boolean => PropValues[PropKeys.TAG].includes(value),
    default: 'span',
  },
  [PropKeys.SIZE]: {
    required: false,
    type: [String, Object] as PropType<PropSize>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.SIZE),
  },
  [PropKeys.FONT]: {
    required: false,
    type: [String, Object] as PropType<PropFont>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.FONT),
  },
  [PropKeys.TRANSFORM]: {
    required: false,
    type: String as PropType<PropTransform>,
    default: 'default',
    ...PropHelpers.stylePropValidation(ConfigStyles, PropKeys.TRANSFORM),
  },
  [PropKeys.LETTER_SPACING]: {
    required: false,
    type: [String, Object] as PropType<PropLetterSpacing>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.LETTER_SPACING),
  },
  [PropKeys.LINE_HEIGHT]: {
    required: false,
    type: [String, Object] as PropType<PropLineHeight>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.LINE_HEIGHT),
  },
  [PropKeys.ALIGN]: {
    required: false,
    type: [String, Object] as PropType<PropAlign>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.ALIGN),
  },
  ...config.props,
};
