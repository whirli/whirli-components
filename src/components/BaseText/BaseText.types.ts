import { PropKeys, PropValues } from './BaseText.constants';

const propColor = PropValues[PropKeys.COLOR];
const propSize = PropValues[PropKeys.SIZE];
const propTag = PropValues[PropKeys.TAG];
const propFamily = PropValues[PropKeys.FAMILY];

export type PropColor = typeof propColor[number];
export type PropSize = typeof propSize[number];
export type PropTag = typeof propTag[number];
export type PropFamily = typeof propFamily[number];
