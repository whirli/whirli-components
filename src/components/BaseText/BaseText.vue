<template>
  <component :is="props.tag" :class="classes"><slot /></component>
</template>

<script setup lang="ts">
// Vue
// Styles
import styles from '@whirli-local/components/BaseText/BaseText.module.scss?module';

// Data
import { ConfigStyles, ConfigProps } from './BaseText.config';

const ComponentStyles = ConfigStyles;

const props = defineProps(ConfigProps);

// Classes
import useClasses from '@whirli-components/@use/class';
const { makeClasses } = useClasses();
const classes = [styles.text, ...makeClasses(ComponentStyles, props, styles)];
</script>
