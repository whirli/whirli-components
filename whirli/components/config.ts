import BaseAside from './BaseAside/BaseAside.config';
import BaseButton from './BaseButton/BaseButton.config';
import BaseColumn from './BaseColumn/BaseColumn.config';
import BaseContainer from './BaseContainer/BaseContainer.config';
import BaseGrid from './BaseGrid/BaseGrid.config';
import BaseImage from './BaseImage/BaseImage.config';
import BaseMain from './BaseMain/BaseMain.config';
import BaseSection from './BaseSection/BaseSection.config';
import BaseSpacer from './BaseSpacer/BaseSpacer.config';
import BaseText from './BaseText/BaseText.config';

export default {
  BaseAside,
  BaseButton,
  BaseColumn,
  BaseContainer,
  BaseGrid,
  BaseImage,
  BaseMain,
  BaseSection,
  BaseSpacer,
  BaseText,
};
