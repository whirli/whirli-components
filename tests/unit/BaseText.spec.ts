import { shallowMount } from '@vue/test-utils';
import { propReturnsClass, propReturnsBreakpointClasses } from '../helpers';

import BaseText from '@/components/BaseText/BaseText.vue';
import { PropKeys } from '@/components/BaseText/BaseText.constants';
import { ConfigStyles } from '@/components/BaseText/BaseText.config';

let wrapper: any;
const text = 'Hello World';

describe('BaseText.vue - Single classes', () => {
  beforeEach(() => {
    wrapper = shallowMount(BaseText, {
      slots: {
        default: text,
      },
      props: {
        [PropKeys.TAG]: 'h1',
        [PropKeys.FONT]: 'bold',
        [PropKeys.SIZE]: 'xl',
        [PropKeys.LINE_HEIGHT]: 'xs',
        [PropKeys.LETTER_SPACING]: 'md',
        [PropKeys.TRANSFORM]: 'uppercase',
        [PropKeys.ALIGN]: 'center',
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('displays the correct text', () => {
    expect(wrapper.text()).toMatch(text);
  });

  it('displays the correct tag', () => {
    expect(wrapper.element.tagName).toBe('H1');
  });

  it('maps to the correct styles', () => {
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.FONT, 'font-bold')).toBe(true);
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.COLOR, 'color-inherit')).toBe(true);
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.SIZE, 'size-xl')).toBe(true);
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.LINE_HEIGHT, 'lineHeight-xs')).toBe(true);
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.LETTER_SPACING, 'letterSpacing-md')).toBe(true);
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.TRANSFORM, 'transform-uppercase')).toBe(true);
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.ALIGN, 'align-center')).toBe(true);
  });
});

describe('BaseText.vue - Breakpoint classes', () => {
  beforeEach(() => {
    wrapper = shallowMount(BaseText, {
      slots: {
        default: text,
      },
      props: {
        [PropKeys.TAG]: 'p',
        [PropKeys.FONT]: 'bold',
        [PropKeys.SIZE]: { default: 'xl', tablet: '3xl' },
        [PropKeys.LINE_HEIGHT]: { default: 'xs', tablet: 'md' },
        [PropKeys.LETTER_SPACING]: { default: 'md', tablet: 'lg' },
        [PropKeys.TRANSFORM]: 'uppercase',
        [PropKeys.ALIGN]: { default: 'center', tablet: 'default' },
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('displays the correct text', () => {
    expect(wrapper.text()).toMatch(text);
  });

  it('displays the correct tag', () => {
    expect(wrapper.element.tagName).toBe('P');
  });

  it('maps to the correct styles', () => {
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.FONT, 'font-bold')).toBe(true);
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.COLOR, 'color-inherit')).toBe(true);
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.TRANSFORM, 'transform-uppercase')).toBe(true);

    expect(
      propReturnsBreakpointClasses(wrapper, ConfigStyles, PropKeys.SIZE, {
        default: 'size-xl',
        tablet: 'size-3xl',
      })
    ).toBe(true);

    expect(
      propReturnsBreakpointClasses(wrapper, ConfigStyles, PropKeys.LINE_HEIGHT, {
        default: 'line-height-xs',
        tablet: 'line-height-md',
      })
    ).toBe(true);

    expect(
      propReturnsBreakpointClasses(wrapper, ConfigStyles, PropKeys.LETTER_SPACING, {
        default: 'letter-spacing-md',
        tablet: 'letter-spacing-lg',
      })
    ).toBe(true);

    expect(
      propReturnsBreakpointClasses(wrapper, ConfigStyles, PropKeys.ALIGN, {
        default: 'align-center',
        tablet: 'align-left',
      })
    ).toBe(true);
  });
});
