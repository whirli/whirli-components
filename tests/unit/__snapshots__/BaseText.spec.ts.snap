// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseText.vue - Breakpoint classes renders the component 1`] = `
<p
  class=""
>
  
  Hello World
  
</p>
`;

exports[`BaseText.vue - Single classes renders the component 1`] = `
<h1
  class=""
>
  
  Hello World
  
</h1>
`;
